executor/stale_txn_test.go

```diff
@@ -291,31 +292,54 @@ func (s *testStaleTxnSuite) TestStalenessAndHistoryRead(c *C) {
 	ON DUPLICATE KEY
 	UPDATE variable_value = '%[2]s', comment = '%[3]s'`, safePointName, safePointValue, safePointComment)
 	tk.MustExec(updateSafePoint)
-	// set @@tidb_snapshot before staleness txn
-	tk.MustExec(`START TRANSACTION READ ONLY AS OF TIMESTAMP '2020-09-06 00:00:00';`)
-	// 1599321600000 == 2020-09-06 00:00:00
-	c.Assert(oracle.ExtractPhysical(tk.Se.GetSessionVars().TxnCtx.StartTS), Equals, int64(1599321600000))
-	tk.MustExec("commit")
-	// set @@tidb_snapshot during staleness txn
-	tk.MustExec(`START TRANSACTION READ ONLY AS OF TIMESTAMP '2020-09-06 00:00:00';`)
-	tk.MustExec(`set @@tidb_snapshot="2016-10-08 16:45:26";`)
-	c.Assert(oracle.ExtractPhysical(tk.Se.GetSessionVars().TxnCtx.StartTS), Equals, int64(1599321600000))
-	tk.MustExec("commit")
 
-	// test mutex
+	// test set txn as of will flush/mutex tidb_snapshot
 	tk.MustExec(`set @@tidb_snapshot="2020-10-08 16:45:26";`)
 	c.Assert(tk.Se.GetSessionVars().SnapshotTS, Equals, uint64(419993151340544000))
 	c.Assert(tk.Se.GetSessionVars().SnapshotInfoschema, NotNil)
 	tk.MustExec("SET TRANSACTION READ ONLY AS OF TIMESTAMP '2020-10-08 16:46:26'")
 	c.Assert(tk.Se.GetSessionVars().SnapshotTS, Equals, uint64(0))
 	c.Assert(tk.Se.GetSessionVars().TxnReadTS.PeakTxnReadTS(), Equals, uint64(419993167069184000))
 
+	// test tidb_snapshot will flush/mutex set txn as of
 	tk.MustExec("SET TRANSACTION READ ONLY AS OF TIMESTAMP '2020-10-08 16:46:26'")
 	c.Assert(tk.Se.GetSessionVars().TxnReadTS.PeakTxnReadTS(), Equals, uint64(419993167069184000))
 	tk.MustExec(`set @@tidb_snapshot="2020-10-08 16:45:26";`)
 	c.Assert(tk.Se.GetSessionVars().SnapshotTS, Equals, uint64(419993151340544000))
+	c.Assert(tk.Se.GetSessionVars().SnapshotInfoschema, NotNil)
 	c.Assert(tk.Se.GetSessionVars().TxnReadTS.PeakTxnReadTS(), Equals, uint64(0))
+
+	c.Assert(tk.Se.GetSessionVars().SetSystemVar(variable.TiDBTxnReadTS, ""), IsNil)
+
+	// test start txn will flush/mutex tidb_snapshot
+	tk.MustExec(`set @@tidb_snapshot="2020-10-08 16:45:26";`)
+	c.Assert(tk.Se.GetSessionVars().SnapshotTS, Equals, uint64(419993151340544000))
```

Comment:
I think we can set `tidb_snapshot`/ `set transaction read only as of timestamp` / `start transaction read only as of timestamp` with different timestamp ( ddl operation between each timestamp) and assert the SchemaVersion of SnapshotInfoSchema.

Reply:
What about now?